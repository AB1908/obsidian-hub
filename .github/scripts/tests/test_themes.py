import os

import approvaltests
from approvaltests import verify_as_json, verify, Options
from approvaltests.scrubbers import create_regex_scrubber, Scrubber
from approvaltests.storyboard import StoryBoard
from jinja2 import Template

import utils
from tests.helpers_for_testing import verify_as_markdown, get_saved_sample_data_for_theme
from tests.test_templates import JINJA_TEMPLATES_DIR
from themes import get_theme_downloads, ThemeList, collect_data_for_theme_and_css
from utils import THEMES_JSON_FILE, get_json_from_github


# TODO  Several of these tests download data from GitHub.
#       This makes the tests slow and prone to failure when data changes.
#       Change them to read test data files from the repo.


def test_get_community_themes() -> None:
    theme_list: ThemeList = get_json_from_github(THEMES_JSON_FILE)
    assert len(theme_list) >= 115, len(theme_list)

    first_theme = theme_list[0]
    assert len(first_theme["modes"]) > 0
    assert first_theme["author"] != ""
    assert first_theme["name"] != ""
    assert first_theme["screenshot"] != ""

    # TODO Don't add new themes - maybe get a fixed revision
    verify_as_json(theme_list)


def test_collect_data_for_theme_with_settings() -> None:
    theme_name = "Minimal"
    verify_theme_data(theme_name)


def test_collect_data_for_theme_without_settings() -> None:
    theme_name = "Christmas"
    verify_theme_data(theme_name)


def test_rendering_of_theme() -> None:
    theme_name = "Minimal"

    template = get_template_for_theme()

    theme, css_file, theme_downloads = get_saved_sample_data_for_theme(theme_name)

    name = collect_data_for_theme_and_css(theme, css_file, theme_downloads, template)
    assert name == theme_name

    file_path = "delete_me.md"
    absolute_file_path = os.path.abspath(file_path)
    file_content = utils.render_template_for_file(template, absolute_file_path, **theme)

    assert len(file_content) > 0

    verify_as_markdown(file_content)


def get_template_for_theme() -> Template:
    template = utils.get_template_from_directory(JINJA_TEMPLATES_DIR, "theme.md.jinja")
    return template


def verify_theme_data(theme_name: str) -> None:
    s = StoryBoard()

    template = get_template_for_theme()

    theme, css_file, theme_downloads = get_saved_sample_data_for_theme(theme_name)

    assert theme
    assert theme["name"] == theme_name
    assert len(theme["modes"]) > 0
    assert theme["author"] != ""

    s.add_frame(approvaltests.utils.to_json(theme))

    name = collect_data_for_theme_and_css(theme, css_file, theme_downloads, template)
    assert name == theme_name
    assert theme["user"] != ""
    s.add_frame(approvaltests.utils.to_json(theme))

    verify(s)
