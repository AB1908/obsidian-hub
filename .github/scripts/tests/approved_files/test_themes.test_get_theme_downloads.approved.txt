{
    "80s Neon": {
        "download": 9999,
        "id": "80s Neon"
    },
    "Agora": {
        "download": 9999,
        "id": "Agora"
    },
    "Al Dente": {
        "download": 9999,
        "id": "Al Dente"
    },
    "Alchemy": {
        "download": 9999,
        "id": "Alchemy"
    },
    "Amethyst": {
        "download": 9999,
        "id": "Amethyst"
    },
    "Ars Magna": {
        "download": 9999,
        "id": "Ars Magna"
    },
    "Atom": {
        "download": 9999,
        "id": "Atom"
    },
    "Aurora": {
        "download": 9999,
        "id": "Aurora"
    },
    "Ayu": {
        "download": 9999,
        "id": "Ayu"
    },
    "Ayu Mirage": {
        "download": 9999,
        "id": "Ayu Mirage"
    },
    "Base2Tone": {
        "download": 9999,
        "id": "Base2Tone"
    },
    "Behave dark": {
        "download": 9999,
        "id": "Behave dark"
    },
    "Big Sur Aesthetic": {
        "download": 9999,
        "id": "Big Sur Aesthetic"
    },
    "Blackbird": {
        "download": 9999,
        "id": "Blackbird"
    },
    "Blue Topaz": {
        "download": 9999,
        "id": "Blue Topaz"
    },
    "Bubble Space": {
        "download": 9999,
        "id": "Bubble Space"
    },
    "California Coast": {
        "download": 9999,
        "id": "California Coast"
    },
    "Cardstock": {
        "download": 9999,
        "id": "Cardstock"
    },
    "Carnelian": {
        "download": 9999,
        "id": "Carnelian"
    },
    "Carpe Noctem": {
        "download": 9999,
        "id": "Carpe Noctem"
    },
    "Charcoal": {
        "download": 9999,
        "id": "Charcoal"
    },
    "Christmas": {
        "download": 9999,
        "id": "Christmas"
    },
    "Cidreira Coast": {
        "download": 9999,
        "id": "Cidreira Coast"
    },
    "Clair de Lune": {
        "download": 9999,
        "id": "Clair de Lune"
    },
    "Comfort Smooth": {
        "download": 9999,
        "id": "Comfort Smooth"
    },
    "Comfort color dark": {
        "download": 9999,
        "id": "Comfort color dark"
    },
    "Creature": {
        "download": 9999,
        "id": "Creature"
    },
    "Cyber Glow": {
        "download": 9999,
        "id": "Cyber Glow"
    },
    "Cybertron": {
        "download": 9999,
        "id": "Cybertron"
    },
    "Dark Graphite": {
        "download": 9999,
        "id": "Dark Graphite"
    },
    "Dark Graphite Pie": {
        "download": 9999,
        "id": "Dark Graphite Pie"
    },
    "Dark Moss": {
        "download": 9999,
        "id": "Dark Moss"
    },
    "Darkyan": {
        "download": 9999,
        "id": "Darkyan"
    },
    "Deep": {
        "download": 9999,
        "id": "Deep"
    },
    "Deep Work": {
        "download": 9999,
        "id": "Deep Work"
    },
    "Discordian": {
        "download": 9999,
        "id": "Discordian"
    },
    "Dracula Official": {
        "download": 9999,
        "id": "Dracula Official"
    },
    "Dracula Slim": {
        "download": 9999,
        "id": "Dracula Slim"
    },
    "Dracula for Obsidian": {
        "download": 9999,
        "id": "Dracula for Obsidian"
    },
    "Ebullientworks": {
        "download": 9999,
        "id": "Ebullientworks"
    },
    "Emerald": {
        "download": 9999,
        "id": "Emerald"
    },
    "Everforest": {
        "download": 9999,
        "id": "Everforest"
    },
    "Faded": {
        "download": 9999,
        "id": "Faded"
    },
    "Firefly": {
        "download": 9999,
        "id": "Firefly"
    },
    "Friendly": {
        "download": 9999,
        "id": "Friendly"
    },
    "GDCT": {
        "download": 9999,
        "id": "GDCT"
    },
    "GDCT Dark": {
        "download": 9999,
        "id": "GDCT Dark"
    },
    "Gastown": {
        "download": 9999,
        "id": "Gastown"
    },
    "Gitsidian": {
        "download": 9999,
        "id": "Gitsidian"
    },
    "Golden Book": {
        "download": 9999,
        "id": "Golden Book"
    },
    "Golden Coffee": {
        "download": 9999,
        "id": "Golden Coffee"
    },
    "Golden Topaz": {
        "download": 9999,
        "id": "Golden Topaz"
    },
    "Harmonic": {
        "download": 9999,
        "id": "Harmonic"
    },
    "Higlighter": {
        "download": 9999,
        "id": "Higlighter"
    },
    "Hipstersmoothie": {
        "download": 9999,
        "id": "Hipstersmoothie"
    },
    "Horizon": {
        "download": 9999,
        "id": "Horizon"
    },
    "Hulk": {
        "download": 9999,
        "id": "Hulk"
    },
    "ITS Theme": {
        "download": 9999,
        "id": "ITS Theme"
    },
    "Iceberg": {
        "download": 9999,
        "id": "Iceberg"
    },
    "Illusion": {
        "download": 9999,
        "id": "Illusion"
    },
    "LYT Mode": {
        "download": 9999,
        "id": "LYT Mode"
    },
    "Lizardmen Zettelkasten": {
        "download": 9999,
        "id": "Lizardmen Zettelkasten"
    },
    "Mammoth": {
        "download": 9999,
        "id": "Mammoth"
    },
    "Minimal": {
        "download": 9999,
        "id": "Minimal"
    },
    "Molecule": {
        "download": 9999,
        "id": "Molecule"
    },
    "Moonlight": {
        "download": 9999,
        "id": "Moonlight"
    },
    "Nebula": {
        "download": 9999,
        "id": "Nebula"
    },
    "Night Owl": {
        "download": 9999,
        "id": "Night Owl"
    },
    "Notation": {
        "download": 9999,
        "id": "Notation"
    },
    "Obsdn-Dark-Rmx": {
        "download": 9999,
        "id": "Obsdn-Dark-Rmx"
    },
    "Obsidian Boom": {
        "download": 9999,
        "id": "Obsidian Boom"
    },
    "Obsidian Nord": {
        "download": 9999,
        "id": "Obsidian Nord"
    },
    "Obsidian Nord Enhanced": {
        "download": 9999,
        "id": "Obsidian Nord Enhanced"
    },
    "Obsidian Windows 98 Edition": {
        "download": 9999,
        "id": "Obsidian Windows 98 Edition"
    },
    "Obsidian You": {
        "download": 9999,
        "id": "Obsidian You"
    },
    "Obsidian gruvbox": {
        "download": 9999,
        "id": "Obsidian gruvbox"
    },
    "Obsidianite": {
        "download": 9999,
        "id": "Obsidianite"
    },
    "Obuntu": {
        "download": 9999,
        "id": "Obuntu"
    },
    "Ono Sendai": {
        "download": 9999,
        "id": "Ono Sendai"
    },
    "Palatinate": {
        "download": 9999,
        "id": "Palatinate"
    },
    "Panic Mode": {
        "download": 9999,
        "id": "Panic Mode"
    },
    "Pine Forest Berry": {
        "download": 9999,
        "id": "Pine Forest Berry"
    },
    "Pink Topaz": {
        "download": 9999,
        "id": "Pink Topaz"
    },
    "Pisum": {
        "download": 9999,
        "id": "Pisum"
    },
    "Primary": {
        "download": 9999,
        "id": "Primary"
    },
    "Prism": {
        "download": 9999,
        "id": "Prism"
    },
    "Purple Aurora": {
        "download": 9999,
        "id": "Purple Aurora"
    },
    "Purple Owl": {
        "download": 9999,
        "id": "Purple Owl"
    },
    "Red Graphite": {
        "download": 9999,
        "id": "Red Graphite"
    },
    "Reverie": {
        "download": 9999,
        "id": "Reverie"
    },
    "Rmaki": {
        "download": 9999,
        "id": "Rmaki"
    },
    "Ros\u00e9 Pine Moon": {
        "download": 9999,
        "id": "Ros\u00e9 Pine Moon"
    },
    "Ruby": {
        "download": 9999,
        "id": "Ruby"
    },
    "Sanctum": {
        "download": 9999,
        "id": "Sanctum"
    },
    "Sandstorm": {
        "download": 9999,
        "id": "Sandstorm"
    },
    "Shimmering Focus": {
        "download": 9999,
        "id": "Shimmering Focus"
    },
    "Slytherin": {
        "download": 9999,
        "id": "Slytherin"
    },
    "Sodalite": {
        "download": 9999,
        "id": "Sodalite"
    },
    "Solarized": {
        "download": 9999,
        "id": "Solarized"
    },
    "Spectrum": {
        "download": 9999,
        "id": "Spectrum"
    },
    "Subtlegold": {
        "download": 9999,
        "id": "Subtlegold"
    },
    "Suddha": {
        "download": 9999,
        "id": "Suddha"
    },
    "Terminal": {
        "download": 9999,
        "id": "Terminal"
    },
    "Things": {
        "download": 9999,
        "id": "Things"
    },
    "Tokyo Night": {
        "download": 9999,
        "id": "Tokyo Night"
    },
    "Traffic Lights": {
        "download": 9999,
        "id": "Traffic Lights"
    },
    "Typewriter": {
        "download": 9999,
        "id": "Typewriter"
    },
    "Typomagical": {
        "download": 9999,
        "id": "Typomagical"
    },
    "Ursa": {
        "download": 9999,
        "id": "Ursa"
    },
    "Vileplume": {
        "download": 9999,
        "id": "Vileplume"
    },
    "Vintage Theme": {
        "download": 9999,
        "id": "Vintage Theme"
    },
    "Warmth": {
        "download": 9999,
        "id": "Warmth"
    },
    "Wasp": {
        "download": 9999,
        "id": "Wasp"
    },
    "WilcoxOne": {
        "download": 9999,
        "id": "WilcoxOne"
    },
    "Wyrd": {
        "download": 9999,
        "id": "Wyrd"
    },
    "Yin and Yang": {
        "download": 9999,
        "id": "Yin and Yang"
    },
    "obsidian_ia": {
        "download": 9999,
        "id": "obsidian_ia"
    }
}
